rust-runtime/aws-smithy-async/src/future/fn_stream/collect.rs

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Module to extend the functionality of `FnStream` to allow for collecting elements of the stream
+//! into collection.
+//!
+//! Majority of the code is borrowed from
+//! <https://github.com/tokio-rs/tokio/blob/fc9518b62714daac9a38b46c698b94ac5d5b1ca2/tokio-stream/src/stream_ext/collect.rs>
+
+/// A trait that signifies that elements can be collected into `T`.
+///
+/// Currently the trait may not be implemented by clients so we can make changes in the future
+/// without breaking code depending on it.
+pub trait Collectable<T>: sealed::CollectablePrivate<T> {}
+
+pub(crate) mod sealed {
+    #[doc(hidden)]
+    pub trait CollectablePrivate<T> {
+        type Collection;
+
+        fn initialize() -> Self::Collection;
+
+        fn extend(collection: &mut Self::Collection, item: T) -> bool;
+
+        fn finalize(collection: &mut Self::Collection) -> Self;
```

Comment:
Why not just take `self` directly instead of `&mut Self`? It doesn't look like this trait needs object safety.

Reply:
As discussed offline `extend` can keep `&mut Self::Collection` for the convenience of the call sites. We can let `finalize` take an owned `Self::Collection` so we can get rid of `std::mem::replace` in the implementation of `finalize` in a consuming type.

Addressed in 5698ee3.